include/libint2/dfbs_generator.h

```diff
@@ -0,0 +1,301 @@
+            for (auto atom: atoms) {
+                auto atom_bs = BasisSet(obs_name_, {atom});
+                obs_shell_vec.emplace_back(atom_bs.shells());
+            }
+            // get primitive shells from AO functions
+            for (auto obs_shells: obs_shell_vec) {
+                primitive_cluster.emplace_back(uncontract(obs_shells));
+            }
+
+            //compute candidate shells
+            candidate_shells_ = datail::candidate_functions(primitive_cluster);
+            cholesky_threshold_ = cholesky_thershold;
+        }
+
+        /// @brief constructor for DFBS generator class, generates density fitting basis set from products of AO shells provided by user
+        /// @param cluster vector of vector of shells for each atom
+        /// @param cholesky_threshold threshold for choosing a product functions via pivoted Cholesky decomposition
+        DFBasisSetGenerator(std::vector<std::vector<Shell>> cluster, const double cholesky_thershold = 1e-7) {
+            std::vector<std::vector<Shell>> primitive_cluster;
+            for (auto i = 0; i < cluster.size(); ++i) {
+                primitive_cluster.emplace_back(uncontract(cluster[i]));
+            }
+            candidate_shells_ = datail::candidate_functions(primitive_cluster);
+            cholesky_threshold_ = cholesky_thershold;
+        }
+
+        DFBasisSetGenerator() = default;
+
+        ~DFBasisSetGenerator() = default;
+
+        /// @brief returns the candidate shells (full set of product functions)
+        std::vector<std::vector<Shell>> candidate_shells() {
+            return candidate_shells_;
+        }
+
+        /// @brief returns the candidate basis set (full set of product functions)
+        /// @warning generates huge and heavily linearly dependent basis sets
+        BasisSet product_basis(){
+            std::vector<Shell> product_shells;
+            for(auto shells: candidate_shells_){
+                product_shells.insert(product_shells.end(), shells.begin(), shells.end());
+            }
+            return BasisSet(std::move(product_shells));
+        }
+
+        /// @brief returns the candidate shells sorted by angular momentum
+        std::vector<std::vector<std::vector<Shell>>> candidates_splitted_in_L() {
+            std::vector<std::vector<std::vector<Shell>>> sorted_shells;
+            for (auto shells: candidate_shells_) {
+                sorted_shells.push_back(datail::split_by_L(shells));
+            }
+            return sorted_shells;
+        }
+
+        /// @brief returns the reduced shells (reduced set of product functions) computed via pivoted Cholesky decomposition
+        std::vector<std::vector<Shell>> reduced_shells() {
+            if (reduced_shells_.size() != 0)
+                return reduced_shells_;
+            else {
+                auto candidate_splitted_in_L = candidates_splitted_in_L();
+                for (auto i = 0; i < candidate_splitted_in_L.size(); ++i) {
```

Comment:
`auto i = 0; i<v.size();...` will produce lots of warnings due to `I` being deduced to be a signed int whereas `vector::size()` returning `size_t` (which is _unsigned_ and `long` or `long long`). This is one of few cases where you want to specify the type explicitly.

here and elsewhere